lxa_iobus/cli/optick.py

```diff
@@ -0,0 +1,501 @@
+    print()
+
+    for node in sorted(nodes["result"]):
+        if not node.startswith("Optick-"):
+            continue
+
+        print(f"  - {node}")
+
+
+async def cmd_calibrate(args):
+    node = args.node
+    input_port = args.input
+    output_port = args.output
+    base_url = args.base_url
+    changeover_delay = args.changeover_delay
+    rounds = args.rounds
+    measurements_per_round = args.measurements_per_round
+
+    input = f"ADC{input_port}"
+    output = f"OUT{output_port}"
+
+    node = await LxaRemoteNode.new(base_url, node)
+
+    adc = node.od.adc
+    outputs = node.od.outputs
+
+    adc_when_low = list()
+    adc_when_high = list()
+
+    for _outer in range(rounds):
+        await outputs.set_low(output)
+
+        # Give the output state some time to propagate
+        time.sleep(changeover_delay)
+
+        for _inner in range(measurements_per_round):
+            val = await adc.read(input)
+            adc_when_low.append(val / 3.3)
+
+        # Give the output state some time to propagate
+        await outputs.set_high(output)
+
+        time.sleep(changeover_delay)
+
+        for _inner in range(measurements_per_round):
+            val = await adc.read(input)
+            adc_when_high.append(val / 3.3)
+
+    await node.close()
+
+    low_mean, low_std = mean_std(adc_when_low)
+    high_mean, high_std = mean_std(adc_when_high)
+
+    if low_mean > high_mean:
+        low_mean, low_std, high_mean, high_std = (
+            high_mean,
+            high_std,
+            low_mean,
+            low_std,
+        )
+
+    # Set the decision boundary at the center point between the average
+    # measurements for high and low values.
+    mid = (low_mean + high_mean) / 2
+
+    # Approximate the distribution of the measurements using a gaussian
+    # distribution ...
+    mid_low_norm = (mid - low_mean) / low_std
+    mid_high_norm = (high_mean - mid) / high_std
+
+    # ... and give a measure of how good the decision boundary is based
+    # on that.
+    low_correct = math.erf(mid_low_norm)
+    high_correct = math.erf(mid_high_norm)
+
+    print()
+    print(f"Trigger level {mid:0.2f} provides:")
+    print(f"  - {int(low_correct * 100)}% correct low estimates")
+    print(f"  - {int(high_correct * 100)}% correct high estimates")
+    print()
+
+    if mid < 0.05:
+        print("WARNING: The trigger level is very low.")
+        print("Are all connections okay and the display set to its maximum brightness?")
+        print()
+
+    if mid > 0.95:
+        print("WARNING: The trigger level is very high.")
+        print("Is the sensor protected from light sources like e.g. direct sunlight?")
+        print()
+
+    if low_correct < 0.95 or high_correct < 0.95:
+        print("WARNING: The trigger level has bad separation")
+        print("Are all connections good, no stray light entering the setup")
+        print("and automatic brightness adjustments disabled as much as possuble?")
```

Comment:
Reviewing this part may be a good fit for @ukleinek, who has a knack for math stuff.

Reply:
Yes, sure. I am not sure if the percent values I print here are statistically sound, but they have been helpful enough, so I think we are good here.